src/Clay/Grid.hs

```diff
@@ -1,58 +1,249 @@
+columnGap :: Size a -> Css
+columnGap = key "column-gap" <> key "grid-column-gap"
+
+-------------------------------------------------------------------------------
+
+-- | Property defines the line names and track sizing functions of the grid rows.
+gridTemplateRows :: GridTrackList a -> Css
+gridTemplateRows = key "grid-template-rows"
+
+-- | Property defines the line names and track sizing functions of the grid columns.
+gridTemplateColumns :: GridTrackList a -> Css
+gridTemplateColumns = key "grid-template-columns"
+
+newtype GridTrackList a = GridTrackList Value
+  deriving (Val, None, Inherit, Initial, Unset)
+
+-- | Create a GridTrackList from a list of sizes
+mkGridTrackList :: [Size a] -> GridTrackList a
+mkGridTrackList = GridTrackList . noCommas
+
+-------------------------------------------------------------------------------
+
+-- | Property defines the line names and track sizing functions of the grid rows.
+gridAutoRows :: GridAutoTrackList a -> Css
+gridAutoRows = key "grid-auto-rows"
 
 -- | Property defines the line names and track sizing functions of the grid columns.
-gridTemplateColumns :: [Size a] -> Css
-gridTemplateColumns = key "grid-template-columns" . noCommas
+gridAutoColumns :: GridAutoTrackList a -> Css
+gridAutoColumns = key "grid-auto-columns"
+
+newtype GridAutoTrackList a = GridAutoTrackList Value
+  deriving (Val, Auto, MinContent, MaxContent, Inherit, Initial, Unset)
+
+mkGridAutoTrackList :: [Size a] -> GridAutoTrackList
+mkGridAutoTrackList = GridAutoTrackList . noCommas
+
+-------------------------------------------------------------------------------
+gridAutoFlow :: GridAutoFlow -> Css
+gridAutoFlow = key "grid-auto-flow"
+
+newtype GridAutoFlow = GridAutoFlow Value
+  deriving (Val, Row, Column, Inherit, Initial, Unset)
+
+dense :: GridAutoFlow
+dense = GridAutoFlow "dense"
+
+rowDense :: GridAutoFlow
+rowDense = GridAutoFlow "row dense"
+
+columnDense :: GridAutoFlow
+columnDense = GridAutoFlow "column dense"
+-------------------------------------------------------------------------------
+-- | Property defines the element location inside grid template
+gridArea :: GridArea -> Css
+gridArea = key "grid-area"
+
+blankGridArea :: GridArea
+blankGridArea = GridArea "."
+
+newtype GridArea = GridArea Text
+  deriving (IsString, Val)
+
+-------------------------------------------------------------------------------
+
+gridRowStart :: GridLocation -> Css
+gridRowStart = key "grid-row-start"
+
+gridRowEnd :: GridLocation -> Css
+gridRowEnd = key "grid-row-end"
+
+gridColumnStart :: GridLocation -> Css
+gridColumnStart = key "grid-column-start"
+
+gridColumnEnd :: GridLocation -> Css
+gridColumnEnd = key "grid-column-end"
+
+gridLocation :: IsSpan -> These Integer GridArea -> GridLocation
+gridLocation isSpan = GridLocationData . MkGridLocationData isSpan
+
+data IsSpan = Span | NoSpan
+  deriving (Show, Eq)
+
+data GridLocation
```

Comment:
Since this is not a known CSS concept (as far as I know), describe a bit what it is and what it is good for.